src/Interpreters/Aggregator.cpp

```diff
@@ -409,6 +412,9 @@ Block Aggregator::Params::getHeader(
 
             res.insert({ type, aggregate.column_name });
         }
+
+        if (grouping_sets_params.isValid())
+            res.insert({ std::make_shared<DataTypeUInt64>(), "__grouping_set" });
```

Comment:
I really think it would be better to add this column after `getHeader` whenever it is needed.
Especially here we pass `grouping_sets_params` only to check if it `isValid`

Reply:
`getHeader` function is used in a lot of different places. I'd prefer to add this column to the header here. Now it passes only `has_grouping_sets` flag, so I don't see a big problem here.